x-pack/plugins/security/server/session_management/session_index.ts

```diff
@@ -0,0 +1,386 @@
+   * @param sessionValue Session index value.
+   */
+  async create(sessionValue: Readonly<Omit<SessionIndexValue, 'metadata'>>) {
+    if (this.indexInitialization) {
+      this.#options.logger.warn(
+        'Attempted to create a new session while session index is initializing.'
+      );
+      await this.indexInitialization;
+    }
+
+    const { sid, ...sessionValueToStore } = sessionValue;
+    try {
+      const {
+        _primary_term: primaryTerm,
+        _seq_no: sequenceNumber,
+      } = await this.#options.clusterClient.callAsInternalUser('create', {
+        id: sid,
+        // We cannot control whether index is created automatically during this operation or not.
+        // But we can reduce probability of getting into a weird state when session is being created
+        // while session index is missing for some reason. This way we'll recreate index with a
+        // proper name and alias. But this will only work if we still have a proper index template.
+        index: SESSION_INDEX_NAME,
+        body: sessionValueToStore,
+        refresh: 'wait_for',
+      });
+
+      return { ...sessionValue, metadata: { primaryTerm, sequenceNumber } } as SessionIndexValue;
+    } catch (err) {
+      this.#options.logger.error(`Failed to create session value: ${err.message}`);
+      throw err;
+    }
+  }
+
+  /**
+   * Re-indexes updated session value.
+   * @param sessionValue Session index value.
+   */
+  async update(sessionValue: Readonly<SessionIndexValue>) {
+    const { sid, metadata, ...sessionValueToStore } = sessionValue;
+    try {
+      const response = await this.#options.clusterClient.callAsInternalUser('index', {
+        id: sid,
+        index: SESSION_INDEX_ALIAS,
+        body: sessionValueToStore,
+        ifSeqNo: metadata.sequenceNumber,
+        ifPrimaryTerm: metadata.primaryTerm,
+        refresh: 'wait_for',
+        ignore: [409],
+      });
+
+      // We don't want to override changes that were made after we fetched session value or
+      // re-create it if has been deleted already. If we detect such a case we discard changes and
+      // return latest copy of the session value instead or `null` if doesn't exist anymore.
+      const sessionIndexValueUpdateConflict = response.status === 409;
+      if (sessionIndexValueUpdateConflict) {
+        this.#options.logger.debug(
+          'Cannot update session value due to conflict, session either do not exist or was already updated.'
+        );
+        return await this.get(sid);
+      }
+
+      return {
+        ...sessionValue,
+        metadata: { primaryTerm: response._primary_term, sequenceNumber: response._seq_no },
+      } as SessionIndexValue;
+    } catch (err) {
+      this.#options.logger.error(`Failed to update session value: ${err.message}`);
+      throw err;
+    }
+  }
+
+  /**
+   * Clears session value with the specified ID. May trigger a removal of other outdated session
+   * values.
```

Comment:
> I'm torn on removing other outdated sessions as part of the user logout flow.

Yeah, same here!

> I can't remember if we discussed this before, but what are your thoughts on using the task manager to periodically perform session cleanup?

Yep, we were briefly mentioned this and it's a good option, I don't have any concerns other than these two:

* I'm not sure if disabling task manager is considered supported scenario (cause security will be disabled as well unless we make it optional dependency and fall back to `deleteByQuery` here which sounds a bit too complex)
* And that means Task Manager will never be able to depend on Security 

The latter sounds acceptable to me, the former may be a breaking change if anyone disabled task_manager in production. What do you think? Is it something we can assume should never happen?

Reply:
> The latter sounds acceptable to me

I agree. If task manager ever requires security, then we can reevaluate this, but I don't foresee this happening in the near future.

> I'm not sure if disabling task manager is considered supported scenario (cause security will be disabled as well unless we make it optional dependency and fall back to deleteByQuery here which sounds a bit too complex)

I talked to Brandon about this one, and we think it's acceptable to add task manager as a required dependency here

Reply:
Good, thanks for confirming!

Reply:
> I agree. If task manager ever requires security, then we can reevaluate this, but I don't foresee this happening in the near future.

This is something we need to keep in mind for audit logging (cc @thomheymann), as we recently discussed auditing the execution of background jobs. Depending on how we implement audit logging, we _may_ need to make task manager depend on the security plugin, instead of the other way around.

Reply:
> we may need to make task manager depend on the security plugin, instead of the other way around.

If we figure out that we need this I think it'd be worth talking to Platform Team - it's not the first time when we are hit by the lack of the support for circular dependencies (the convoluted workaround for Spaces <-> Security dependency comes to my mind). Maybe we can finally figure something out.

Reply:
Yeah I agree. I imagine they'll be hesitant to actually support circular dependencies, but hopefully there are other ideas we can explore.

We could also investigate developing audit logging in its own plugin, rather than as part of the security plugin. One complication that immediately comes to mind is the configuration namespace. In order to maintain BWC, _and_ to be consistent with ES, I'd like to keep the audit logging configuration within the `xpack.security.audit` namespace.

Reply:
> I'd like to keep the audit logging configuration within the xpack.security.audit namespace.

++.

I also agree that having a "direct" support for circular dependencies is unlikely, and as you mentioned I also think the solution somewhere in a separate plugins. _But_ even though it's acceptable for the time being, the current state of the plugin system would make it a bit confusing (since currently plugins are treated as relatively independent things), we need a better way to scale plugins horizontally (or shard them if you will). 

I don't want to re-invent x-pack, but having native support for something like `domains\features` (e.g. `authentication`, `authorization`, `audit`) within plugin that can share config namespace and be a part of dependency declarations could probably help (e.g. `plugin: Security, domain: authentication` depends on `plugin: TaskManager, domain: scheduler` and `plugin: TaskManager, domain: executor` depends on `plugin: Security, domain: audit` would be allowed scenarios since there is no circular dependencies between domains). Basically it's just a bit of sugar on top of the existing plugin system. Just thinking aloud :see_no_evil: